src/DragDropMonitor.js

```diff
@@ -23,9 +23,24 @@ export default class DragDropMonitor {
       'handlerIds, when specified, must be an array of strings.'
     );
 
+    let prevStateId = this.store.getState().stateId;
     const handleChange = () => {
-      if (areDirty(this.store.getState().dirtyHandlerIds, handlerIds)) {
-        listener();
+      const state = this.store.getState();
+      const currentStateId = state.stateId;
+      const expectedStateId = prevStateId + 1
+
+      try {
+        if (currentStateId === expectedStateId) {
+          if (areDirty(state.dirtyHandlerIds, handlerIds)) {
+            listener();
+          }
+        } else if (currentStateId !== prevStateId) {
+          // Redux do not guarantee that all state updates are sent to all subscribers.
+          // Here we missed at least one state update, so areDirty() will not work.
+          listener();
```

Comment:
Nice,

I have updated the PR with the changes above.

/Mattias

On Wed, Jan 27, 2016 at 1:17 AM, Dan Abramov <notifications@github.com>
wrote:

> In src/DragDropMonitor.js
> <https://github.com/gaearon/dnd-core/pull/25#discussion_r50924041>:
>
> >      const handleChange = () => {
> > -      if (areDirty(this.store.getState().dirtyHandlerIds, handlerIds)) {
> > -        listener();
> > +      const state = this.store.getState();
> > +      const currentStateId = state.stateId;
> > +      const expectedStateId = prevStateId + 1
> > +
> > +      try {
> > +        if (currentStateId === expectedStateId) {
> > +          if (areDirty(state.dirtyHandlerIds, handlerIds)) {
> > +            listener();
> > +          }
> > +        } else if (currentStateId !== prevStateId) {
> > +          // Redux do not guarantee that all state updates are sent to all subscribers.
> > +          // Here we missed at least one state update, so areDirty() will not work.
> > +          listener();
>
> Can we refactor this part to something like
>
> const canSkipListener = currentStateId === prevStateId || (
>   currentStateId === prevStateId + 1 &&
>   !areDirty(state.dirtyHandlerIds, handlerIds)
> )
> if (!canSkipListener) {
>   listener();
> }
>
> ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gaearon/dnd-core/pull/25/files#r50924041>.
>